### PR TITLE
[download_dsyms] wait for build to process

### PR DESCRIPTION
**Pull request by tmm1**
When specifying a version/build_number to download dsyms for, wait to make sure the build has finished processing and dsyms are available first.

This makes it much easier to deal with multiple builds, since pilot can be used to upload them back-to-back, and the slow part of waiting for them to process can be deferred allowing all builds to be processed in parallel.

### CODE REVIEW

**Comment by janpio** [comment]
Hm, how does that change existing behavior? 
How many HTTP calls does it add via spaceship, could you please check via a before/after comparison of a spaceship log (`/tmp/spaceship*.log`) please?

**Comment by tmm1** [comment]
The existing behavior is to fail when it attempts to download dSYMs for a build which has not yet been processed (i.e. dSYMs have not been generated yet).

The spaceship log shows requests to `providers/x/apps/y/platforms/z/trains/a/builds`

The number of requests depends on how long Apple takes to process the build.

**Comment by janpio** [comment]
Yeah, I meant the current default use case of downloading a dsym for an app that _is_ processed. Does this add the call for the trains-builds there as well? Can we avoid that?

**Comment by tmm1** [comment]
I can put the wait call behind an option if you prefer. Otherwise there's no way to know if the build is processed or not without making atleast one call to the train.

**Comment by janpio** [comment]
No changes required, I am just trying to understand if the solution is appropriate.

What happens if one tries to download a dsym of an app that is still processing?

**Comment by tmm1** [comment]
There is some sort of error. Likely saying version not found or something similar.

An alternative that would be more flexible would be to add a new action that just waits for a build to process. That way it would be up to the user to compose it as they like.. so they could call it before they call download_dsyms, or use it elsewhere in their pipeline. Right now the only way to wait for processing is as part of `pilot`, which makes it hard to upload multiple builds together to allow them to be processed in parallel.

File: fastlane/lib/fastlane/actions/download_dsyms.rb
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/download_dsyms.rb]: Thoughts on adding a new option around this? Similar to the `pilot` one that is called `skip_waiting_for_build_processing`? Even though I think this is 💯 myself, I don't want to have to _force_ users to wait for processing if they don't want to wait for whatever reason 🙃 

I think I'm more concerned with breaking/changing any existing behavior/implementation around how users are currently expecting this work. I'm okay with making this new option default to `true` but just at least give the user an option to opt out 🤔

**Comment by PoissonBallon** [comment]
Any news about it ? 
I often have `No dSYM files found on App Store Connect - this usually happens when no recompiling has happened yet` when I tried to download dsyms of my last upload :/

**Comment by tmm1** [comment]
I've been having some issues using this locally when I have multiple apps with the same version/build number. Seems like a similar bug to #14033

File: fastlane/lib/fastlane/actions/download_dsyms.rb
Comment (tmm1) [reply fastlane/lib/fastlane/actions/download_dsyms.rb]: I think that makes sense. I ran into some cases where it would watch indefinitely and never find the build due to other bugs.

**Comment by janpio** [comment]
(The Ubuntu CI failure is not your fault: https://github.com/fastlane/fastlane/issues/14481)

File: fastlane/lib/fastlane/actions/download_dsyms.rb
Comment (gregkerzhner) [reply fastlane/lib/fastlane/actions/download_dsyms.rb]: I guess if you didn't want to introduce breaking changes, the best thing to do might be to have it default to not waiting (old behavior), and introduce a new option that turns your feature on only if the user opts in.  I can see how other people might get confused when this feature hits about why their CI is hanging on some task all of a sudden.

**Comment by joshdholtz** [comment]
@tmm1 If you would allow the maintainers access to make changes to your branch that would be 💯 I would be happy to fix this Ubuntu test error and make the changes needed to get this merged 🙃

https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by tmm1** [comment]
Looks enabled by default already. Would be great if you can take this over!

**Comment by PatrickDotStar** [comment]
We have the same issues for our project. We recently enabled bitcode and now when refreshing the dsym after uploading the app to appstore connect it may be the case that the dsym are not available atm even when the build finished processing. We pass the `wait_for_uploaded_build: true` parameter for the `testflight` command. 
I noticed, that the donwload dsym button in appstore connect is not showing immediately after the build finished processing. It can take some time until they are available for download and if this is not the case the `download_dsyms` script will fail saying that no dSYM URL is available and we then have to run the script again at a later moment

missingDsyms

**Comment by tmm1** [comment]
> I noticed, that the download dsym button in appstore connect is not showing immediately after the build finished processing. It can take some time until they are available for download and if this is not the case the `download_dsyms` script will fail saying that no dSYM URL is available and we then have to run the script again at a later moment

I'm noticing this now too. The whole point of this PR was to wait for processing before downloading, but if that is now not enough I'm not sure what to do. We would need to keep polling for the dSYM url I suppose?

**Comment by bguidolim** [comment]
As a workaround, I simply put a `sleep(60)` before calling my refresh dsyms lane.

**Comment by PatrickDotStar** [comment]
> As a workaround, I simply put a `sleep(60)` before calling my refresh dsyms lane.

Haha we also do the same but we sleep for 3 minutes just to be sure😅
60 sec was not enough for us.

**Comment by tmm1** [comment]
> Haha we also do the same but we sleep for 3 minutes just to be sure😅
> 60 sec was not enough for us.

Yikes.

@joshdholtz It seems like the best approach here is to add a loop around the dSYMs URL fetch instead of giving up after the first failure.

**Comment by tmm1** [comment]
@joshdholtz do you have any time this week to take a look here?

**Review by janpio — APPROVED** [review]
Code itself looks good, not sure of the internal logic though - I would defer that to @joshdholtz to evaluate before merging.

**Comment by notahat** [comment]
I've been experimenting around the same problem, and `BuildWatcher.wait_for_build_processing_to_be_complete` is not enough to guarantee that the dSYMs will be available for download. My experience has been that they're usually available within 30 seconds of that.

I'm considering a PR that adds an option to make `download_dsyms` poll until they're ready (with an eventual timeout.)

**Comment by tmm1** [comment]
> I'm considering a PR that adds an option to make `download_dsyms` poll until they're ready (with an eventual timeout.)

Please do this!

**Comment by janpio** [comment]
(Thanks for your continued work on this issue - it's a really hard nut to crack and I hope we can find a good solution with all your help.)

**Pull request closed without merging**
